File: macos/Tests/UsageCoreTests/UsageClientTests.swift
import XCTest
@testable import UsageCore

final class UsageClientTests: XCTestCase {
    private static let sampleJson =
        "{\"five_hour\":{\"utilization\":52.0,\"resets_at\":\"2026-06-10T16:40:00.627509+00:00\"},"
        + "\"seven_day\":{\"utilization\":27.0,\"resets_at\":\"2026-06-13T11:00:00.627529+00:00\"},"
        + "\"seven_day_oauth_apps\":null,\"seven_day_opus\":null,"
        + "\"seven_day_sonnet\":{\"utilization\":0.0,\"resets_at\":null},"
        + "\"seven_day_cowork\":null,\"seven_day_omelette\":null,\"tangelo\":null,"
        + "\"iguana_necktie\":null,\"omelette_promotional\":null,\"cinder_cove\":null,"
        + "\"extra_usage\":{\"is_enabled\":false,\"monthly_limit\":null,\"used_credits\":null,"
        + "\"utilization\":null,\"currency\":null,\"disabled_reason\":null}}"

    /// Closure transport that records the request and returns a canned status/body.
    private func makeClient(
        status: Int,
        body: String,
        captured: CapturedBox = CapturedBox()
    ) -> (UsageClient, CapturedBox) {
        let client = UsageClient(transport: ClosureTransport { request in
            captured.request = request
            return (Data(body.utf8), status)
        })
        return (client, captured)
    }

    func testFetchParsesSampleJson() async throws {
        let (client, _) = makeClient(status: 200, body: Self.sampleJson)
        let snapshot = try await client.fetch(accessToken: "tok")

        XCTAssertNotNil(snapshot.fiveHour)
        XCTAssertEqual(snapshot.fiveHour!.utilization, 52.0)

        let expected = ISO8601DateParsing.parse("2026-06-10T16:40:00.627509+00:00")
        XCTAssertNotNil(expected)
        XCTAssertEqual(snapshot.fiveHour!.resetsAt!.timeIntervalSince1970,
                       expected!.timeIntervalSince1970, accuracy: 0.001)

        XCTAssertNotNil(snapshot.sevenDay)
        XCTAssertEqual(snapshot.sevenDay!.utilization, 27.0)

        XCTAssertNil(snapshot.sevenDayOpus)

        XCTAssertNotNil(snapshot.sevenDaySonnet)
        XCTAssertEqual(snapshot.sevenDaySonnet!.utilization, 0.0)
        XCTAssertNil(snapshot.sevenDaySonnet!.resetsAt)
    }

    func testFetchSendsRequiredHeaders() async throws {
        let (client, captured) = makeClient(status: 200, body: Self.sampleJson)
        _ = try await client.fetch(accessToken: "tok")

        let req = try XCTUnwrap(captured.request)
        XCTAssertEqual(req.value(forHTTPHeaderField: "Authorization"), "Bearer tok")
        XCTAssertEqual(req.value(forHTTPHeaderField: "anthropic-beta"), "oauth-2025-04-20")
        XCTAssertEqual(req.value(forHTTPHeaderField: "User-Agent"), "claude-code/2.1.170")
        XCTAssertTrue(req.value(forHTTPHeaderField: "User-Agent")!.contains("claude-code"))
        XCTAssertEqual(req.value(forHTTPHeaderField: "Accept"), "application/json")
    }

    func testFetch401ThrowsAuthError() async {
        let (client, _) = makeClient(status: 401, body: "{}")
        await assertThrows(client, expecting: .auth(401))
    }

    func testFetch403ThrowsAuthError() async {
        let (client, _) = makeClient(status: 403, body: "{}")
        await assertThrows(client, expecting: .auth(403))
    }

    func testFetch429ThrowsRateLimitError() async {
        let (client, _) = makeClient(status: 429, body: "{}")
        await assertThrows(client, expecting: .rateLimited)
    }

    func testFetch500ThrowsGenericError() async {
        let (client, _) = makeClient(status: 500, body: "{}")
        do {
            _ = try await client.fetch(accessToken: "tok")
            XCTFail("expected error")
        } catch UsageError.generic {
            // expected
        } catch {
            XCTFail("unexpected error: \(error)")
        }
    }

    func testFetchNullWindowProducesAllNullSnapshot() async throws {
        let (client, _) = makeClient(status: 200, body: "{\"five_hour\":null}")
        let snapshot = try await client.fetch(accessToken: "tok")

        XCTAssertNil(snapshot.fiveHour)
        XCTAssertNil(snapshot.sevenDay)
        XCTAssertNil(snapshot.sevenDayOpus)
        XCTAssertNil(snapshot.sevenDaySonnet)
    }

    func testFetchParsesPlainIso8601WithoutFractionalSeconds() async throws {
        let (client, _) = makeClient(
            status: 200,
            body: "{\"five_hour\":{\"utilization\":10.0,\"resets_at\":\"2026-06-10T16:40:00Z\"}}")
        let snapshot = try await client.fetch(accessToken: "tok")
        XCTAssertNotNil(snapshot.fiveHour!.resetsAt)
    }

    // MARK: - helpers

    private func assertThrows(
        _ client: UsageClient,
        expecting expected: UsageError,
        file: StaticString = #filePath,
        line: UInt = #line
    ) async {
        do {
            _ = try await client.fetch(accessToken: "tok")
            XCTFail("expected error", file: file, line: line)
        } catch let error as UsageError {
            XCTAssertEqual(error, expected, file: file, line: line)
        } catch {
            XCTFail("unexpected error: \(error)", file: file, line: line)
        }
    }
}

/// Reference box so the @Sendable closure can write back the captured request.
final class CapturedBox: @unchecked Sendable {
    var request: URLRequest?
}

File: tests/ClaudeUsageWatcher.Core.Tests/UsageClientTests.cs
using System.Net;
using System.Text;
using ClaudeUsageWatcher.Core;
using Xunit;

namespace ClaudeUsageWatcher.Core.Tests;

public class UsageClientTests
{
    private const string SampleJson =
        "{\"five_hour\":{\"utilization\":52.0,\"resets_at\":\"2026-06-10T16:40:00.627509+00:00\"}," +
        "\"seven_day\":{\"utilization\":27.0,\"resets_at\":\"2026-06-13T11:00:00.627529+00:00\"}," +
        "\"seven_day_oauth_apps\":null,\"seven_day_opus\":null," +
        "\"seven_day_sonnet\":{\"utilization\":0.0,\"resets_at\":null}," +
        "\"seven_day_cowork\":null,\"seven_day_omelette\":null,\"tangelo\":null," +
        "\"iguana_necktie\":null,\"omelette_promotional\":null,\"cinder_cove\":null," +
        "\"extra_usage\":{\"is_enabled\":false,\"monthly_limit\":null,\"used_credits\":null," +
        "\"utilization\":null,\"currency\":null,\"disabled_reason\":null}}";

    private sealed class FakeHandler : HttpMessageHandler
    {
        private readonly HttpStatusCode _status;
        private readonly string _body;

        public HttpRequestMessage? CapturedRequest { get; private set; }

        public FakeHandler(HttpStatusCode status, string body)
        {
            _status = status;
            _body = body;
        }

        protected override Task<HttpResponseMessage> SendAsync(
            HttpRequestMessage request, CancellationToken cancellationToken)
        {
            CapturedRequest = request;
            var response = new HttpResponseMessage(_status)
            {
                Content = new StringContent(_body, Encoding.UTF8, "application/json")
            };
            return Task.FromResult(response);
        }
    }

    [Fact]
    public async Task FetchAsync_ParsesSampleJson()
    {
        var handler = new FakeHandler(HttpStatusCode.OK, SampleJson);
        using var http = new HttpClient(handler);
        var client = new UsageClient(http);

        UsageSnapshot snapshot = await client.FetchAsync("tok");

        Assert.NotNull(snapshot.FiveHour);
        Assert.Equal(52.0, snapshot.FiveHour!.Utilization);
        Assert.Equal(
            DateTimeOffset.Parse("2026-06-10T16:40:00.627509+00:00"),
            snapshot.FiveHour.ResetsAt);

        Assert.NotNull(snapshot.SevenDay);
        Assert.Equal(27.0, snapshot.SevenDay!.Utilization);

        Assert.Null(snapshot.SevenDayOpus);

        Assert.NotNull(snapshot.SevenDaySonnet);
        Assert.Equal(0.0, snapshot.SevenDaySonnet!.Utilization);
        Assert.Null(snapshot.SevenDaySonnet.ResetsAt);
    }

    [Fact]
    public async Task FetchAsync_SendsRequiredHeaders()
    {
        var handler = new FakeHandler(HttpStatusCode.OK, SampleJson);
        using var http = new HttpClient(handler);
        var client = new UsageClient(http);

        await client.FetchAsync("tok");

        HttpRequestMessage req = Assert.IsType<HttpRequestMessage>(handler.CapturedRequest);
        Assert.Equal("Bearer", req.Headers.Authorization!.Scheme);
        Assert.Equal("tok", req.Headers.Authorization.Parameter);

        Assert.True(req.Headers.TryGetValues("anthropic-beta", out var beta));
        Assert.Equal("oauth-2025-04-20", Assert.Single(beta!));

        Assert.True(req.Headers.TryGetValues("User-Agent", out var ua));
        Assert.Contains(ua!, v => v.Contains("claude-code"));
    }

    [Fact]
    public async Task FetchAsync_401_ThrowsAuthException()
    {
        var handler = new FakeHandler(HttpStatusCode.Unauthorized, "{}");
        using var http = new HttpClient(handler);
        var client = new UsageClient(http);

        await Assert.ThrowsAsync<UsageAuthException>(() => client.FetchAsync("tok"));
    }

    [Fact]
    public async Task FetchAsync_429_ThrowsRateLimitException()
    {
        var handler = new FakeHandler(HttpStatusCode.TooManyRequests, "{}");
        using var http = new HttpClient(handler);
        var client = new UsageClient(http);

        await Assert.ThrowsAsync<UsageRateLimitException>(() => client.FetchAsync("tok"));
    }

    [Fact]
    public async Task FetchAsync_NullWindow_ProducesAllNullSnapshot()
    {
        var handler = new FakeHandler(HttpStatusCode.OK, "{\"five_hour\":null}");
        using var http = new HttpClient(handler);
        var client = new UsageClient(http);

        UsageSnapshot snapshot = await client.FetchAsync("tok");

        Assert.Null(snapshot.FiveHour);
        Assert.Null(snapshot.SevenDay);
        Assert.Null(snapshot.SevenDayOpus);
        Assert.Null(snapshot.SevenDaySonnet);
    }
}
